src/shared/varlink.c

```diff
@@ -543,9 +599,42 @@ static int varlink_read(Varlink *v) {
 
         bool prefer_read = v->prefer_read_write;
         if (!prefer_read) {
-                n = recv(v->fd, v->input_buffer + v->input_buffer_index + v->input_buffer_size, rs, MSG_DONTWAIT);
+                struct msghdr msgh = {0};
+                struct iovec iov;
+                union {
+                        char buf[CMSG_SPACE(sizeof(struct ucred))];
+                        struct cmsghdr align;
+                } cmsg;
+                struct cmsghdr *cmsgp;
+
+                iov.iov_base = v->input_buffer + v->input_buffer_index + v->input_buffer_size;
+                iov.iov_len = rs;
+                msgh.msg_iov = &iov;
+                msgh.msg_iovlen = 1;
+                msgh.msg_control = cmsg.buf;
+                msgh.msg_controllen = sizeof(cmsg.buf);
+
+                n = recvmsg(v->fd, &msgh, MSG_DONTWAIT);
                 if (n < 0 && errno == ENOTSOCK)
                         prefer_read = v->prefer_read_write = true;
+
+                if (n != -1) {
```

Comment:
in our codebase we usually check `n < 0` only for cases like this, but doesn't matter